python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1107 @@
+        # Get full data shape
+        for name, data in self._data_store.items():
+            data_shape = list(F.shape(data))
+            data_shape[0] = 0
+            request = GetFullShapeRequest(name)
+            # send request
+            for machine_id in range(self._machine_count):
+                server_id = machine_id * self._group_count
+                rpc.send_request(server_id, request)
+            # recv response
+            for _ in range(self._machine_count):
+                response = rpc.recv_response()
+                data_shape[0] += response.shape[0]
+            self._full_data_shape[name] = tuple(data_shape)
+
+    def barrier(self):
+        """Barrier for all client nodes
+
+        This API will be blocked untill all the clients call this API.
+        """
+        request = BarrierRequest(self._client_id)
+        # send request
+        for server_id in range(self._server_count):
+            rpc.send_request(server_id, request)
+        # recv response
+        for _ in range(self._server_count):
+            response = rpc.recv_response()
+
+    def init_data(self, name, shape, dtype, policy_str, partition_book, init_func):
+        """Send message to kvserver to initialize new data and 
+        get corresponded shared-tensor on kvclient. 
+
+        Parameters
+        ----------
+        name : str
+            data name
+        shape : list or tuple of int
+            data shape
+        dtype : dtype
+            data type
+        policy_str : str
+            partition policy string, e.g., 'edge' or 'node'.
+        partition_book : GraphPartitionBook or RangePartitionBook
+            Store the partition information
+        init_func : func
+            UDF init function
+        """
+        assert len(name) > 0, 'name cannot be empty.'
+        assert len(shape) > 0, 'shape cannot be empty'
+        assert policy_str in ('edge', 'node'), 'policy_str must be \'edge\' or \'node\'.'
+        shape = list(shape)
+        policy_list = []
+        if self._client_id == 0:
+            for machine_id in range(self._machine_count):
+            	policy = PartitionPolicy(plolicy_str, machine_id, partition_book)
+                part_shape = shape.copy()
+                part_shape[0] = policy.get_data_size()
+                request = InitDataRequest(name,
+                                          tuple(part_shape), 
+                                          get_type_str(dtype), 
+                                          policy_str,
+                                          init_func)
+                for n in range(self._group_count):
+                    server_id = machine_id * self._group_count + n
+                    rpc.send_request(server_id, request)
+            for _ in range(self._server_count):
+                response = rpc.recv_response()
+            self.barrier() # wait all the client and server finish its job
```

Comment:
do you still need the barrier? you already waited for all servers' response.

Reply:
Just change the logic.